trino-aws-proxy-spark3/src/main/java/io/trino/aws/proxy/spark3/PresignAwareAmazonS3.java

```diff
@@ -0,0 +1,1667 @@
+import com.amazonaws.services.s3.model.VersionListing;
+import com.amazonaws.services.s3.model.WriteGetObjectResponseRequest;
+import com.amazonaws.services.s3.model.WriteGetObjectResponseResult;
+import com.amazonaws.services.s3.model.analytics.AnalyticsConfiguration;
+import com.amazonaws.services.s3.model.intelligenttiering.IntelligentTieringConfiguration;
+import com.amazonaws.services.s3.model.inventory.InventoryConfiguration;
+import com.amazonaws.services.s3.model.metrics.MetricsConfiguration;
+import com.amazonaws.services.s3.model.ownership.OwnershipControls;
+import com.amazonaws.services.s3.waiters.AmazonS3Waiters;
+import com.amazonaws.util.IOUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+class PresignAwareAmazonS3
+        implements AmazonS3
+{
+    private static final Log log = LogFactory.getLog(PresignAwareAmazonS3.class);
+
+    private final AmazonS3 delegate;
+
+    PresignAwareAmazonS3(AmazonS3 delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    private record Presigned(URL url, ObjectMetadata objectMetadata)
+    {
+        private Presigned
+        {
+            requireNonNull(url, "url is null");
+            requireNonNull(objectMetadata, "objectMetadata is null");
+        }
+    }
+
+    private Optional<Presigned> getPresignedUrl(String method, String bucket, String key)
+    {
+        return getPresignedUrl(method, bucket, key, Optional.empty());
+    }
+
+    private Optional<Presigned> getPresignedUrl(String method, String bucket, String key, Optional<UploadPartRequest> uploadPartRequest)
+    {
+        Optional<String> uploadId = uploadPartRequest.map(UploadPartRequest::getUploadId);
+        Optional<Integer> partNumber = uploadPartRequest.map(UploadPartRequest::getPartNumber);
+
+        log.debug("request.get-presigned-url method=%s bucket=%s key=%s uploadId=%s partNumber=%s".formatted(method, bucket, key, uploadId, partNumber));
+
+        Map<String, ?> headers;
+        ObjectMetadata objectMetadata;
+        try {
+            GetObjectMetadataRequest getObjectMetadataRequest = new GetObjectMetadataRequest(bucket, key);
+            partNumber.ifPresent(getObjectMetadataRequest::setPartNumber);
+            uploadId.ifPresent(id -> getObjectMetadataRequest.putCustomQueryParameter("uploadId", id));
+
+            objectMetadata = delegate.getObjectMetadata(getObjectMetadataRequest);
+            headers = objectMetadata.getRawMetadata();
+        }
+        catch (AmazonServiceException e) {
+            headers = e.getHttpHeaders();
```

Comment:
> For such cases can't we just return the equivalent of a 404?

404 _is_ returned. However, the AWS client turns that into an exception.

Reply:
Sorry, I meant return to what is using `PresignAwareAmazonS3` (Spark in this case).

So if this does a HEAD on a bucket/key to get the presgined URL and gets a 404 it shouldn't then try to use the presigned GET/DELETE URL to access that resource as it would again result in a 404. This function can (presumably) just re-throw the `AmazonServiceException`. 

This only applies to GET and DELETE requests (which I forgot to mention in the initial comment) as PUT/POST would create the object if its missing.

Reply:
Oh I see - let me look into that.

Reply:
@mosiac1 OK - I've updated to handle this